segmentstore/server/host/src/test/java/io/pravega/segmentstore/server/host/ExtS3IntegrationTest.java

```diff
@@ -0,0 +1,359 @@
+        baseDir = Files.createTempDirectory("exts3_wrapper").toString();
+
+        this.configBuilder.include(ExtS3StorageConfig.builder()
+                                                     .with(ExtS3StorageConfig.EXTS3_BUCKET, "kanpravegatest")
+                                                     .with(ExtS3StorageConfig.EXTS3_ACCESS_KEY_ID, "x")
+                                                     .with(ExtS3StorageConfig.EXTS3_SECRET_KEY, "x")
+                                                     .with(ExtS3StorageConfig.EXTS3_URI, "http://localhost:9020"));
+    }
+
+    /**
+     * Shuts down BookKeeper and cleans up file system directory
+     */
+    @After
+    public void tearDown() throws Exception {
+        bookkeeper.close();
+        FileHelpers.deleteFileOrDirectory(new File(baseDir));
+    }
+
+    //endregion
+
+    //region StreamSegmentStoreTestBase Implementation
+
+    @Override
+    protected ServiceBuilder createBuilder(ServiceBuilderConfig builderConfig) {
+        return ServiceBuilder
+                .newInMemoryBuilder(builderConfig)
+                .withCacheFactory(setup -> new RocksDBCacheFactory(builderConfig.getConfig(RocksDBConfig::builder)))
+                .withStorageFactory(setup -> {
+                    StorageFactory f = new ExtS3StorageFactory(
+                            setup.getConfig(ExtS3StorageConfig::builder), setup.getExecutor());
+                    return new ExtS3ListenableStorageFactory(f, builderConfig.getConfig(ExtS3StorageConfig::builder));
+                })
+                .withDataLogFactory(setup -> new BookKeeperLogFactory(setup.getConfig(BookKeeperConfig::builder),
+                        bookkeeper.getZkClient(), setup.getExecutor()));
+    }
+
+    private class ExtS3ListenableStorageFactory extends ListenableStorageFactory {
+
+        private final ExtS3StorageConfig adapterConfig;
+
+        public ExtS3ListenableStorageFactory(StorageFactory wrappedFactory, ExtS3StorageConfig adapterConfig) {
+            super(wrappedFactory);
+            this.adapterConfig = adapterConfig;
+        }
+
+        @Override
+        public Storage createStorageAdapter() {
+            S3Config exts3Config = null;
+            try {
+                exts3Config = new S3Config(new URI("http://localhost:9020"));
+            } catch (URISyntaxException e) {
+                throw new IllegalArgumentException("Wrong URI");
+            }
+            exts3Config = exts3Config.withIdentity( adapterConfig.getExts3AccessKey()).withSecretKey( adapterConfig.getExts3SecretKey())
+                                     .withRetryEnabled( false).withInitialRetryDelay( 1).withProperty("com.sun.jersey.client.property.connectTimeout", 100);
+
+            S3JerseyClient client = new S3ClientWrapper(exts3Config, baseDir);
+
+            Storage storage = this.wrappedFactory.createStorageAdapter();
+            this.storage = storage;
+            ((ExtS3Storage) storage).setClient(client);
```

Comment:
This is the whole point of the `StorageFactory`. You should never have to do hacks like this. Please find a way to do this via the factory.